{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(GSVA)\n",
    "library(clusterProfiler)\n",
    "library(survival)\n",
    "library(survminer)\n",
    "library(Publish)\n",
    "library(dplyr)\n",
    "library(readxl)\n",
    "library(ggsci)\n",
    "library(compareGroups)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "setwd(\"/media/bio/Disk/Research Data/EBV/omicverse\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Prepare the bulk data and phenotype\n",
    "GSE102349_Dataset <- read.csv(file=\"Dataset/GEO/GSE102349/02_normalized/GSE102349_normalized.csv\",row.names=1,check.names=F)\n",
    "# Same with the LUAD scRNA-seq data, the rows and columns are genes and cells, respectively. The dimensions of the LUAD bulk dataset are:\n",
    "dim(GSE102349_Dataset)\n",
    "GSE102349_Dataset <- GSE102349_Dataset %>% replace(., is.na(.), 0)\n",
    "# It shows this data has 56716 genes and 471 samples in total. Users do not need to keep the common genes between the single-cell and bulk samples, which can be automatically achieved in Scissor. Besides, all of these samples have clinical outcomes:\n",
    "\n",
    "# The clinical survival phenotype requires a two-column matrix with columns named 'time' and 'status'. The first column contains the sample IDs (observations), which should in the same order as the columns in the bulk expression matrix. The second column is a binary variable, with '1' indicating event (e.g.recurrence of cancer or death), and '0' indicating right-censored.\n",
    "GSE102349_Clinical = read_excel(\"Dataset/GEO/GSE102349/clinical.xls\",sheet = 1) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#GEO ssGSEA\n",
    "Myeloid.Markers <- read.csv(\"Results/07.Myeloid/cosg_marker_gene_leiden_0_2.csv\")\n",
    "T.Markers <- read.csv(\"Results/11.TCell/cosg_marker_gene_leiden_1_0.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " Stem.Sig2 <- c(\n",
    "  \"DNMT3B\", \"PFAS\", \"XRCC5\", \"HAUS6\", \"TET1\", \"IGF2BP1\", \"PLAA\", \"TEX10\", \"MSH6\", \"DLGAP5\",\n",
    "  \"SKIV2L2\", \"SOHLH2\", \"RRAS2\", \"PAICS\", \"CPSF3\", \"LIN28B\", \"IPO5\", \"BMPR1A\", \"ZNF788\", \"ASCC3\",\n",
    "  \"FANCB\", \"HMGA2\", \"TRIM24\", \"ORC1\", \"HDAC2\", \"HESX1\", \"INHBE\", \"MIS18A\", \"DCUN1D5\", \"MRPL3\", \"CENPH\",\n",
    "  \"MYCN\", \"HAUS1\", \"GDF3\", \"TBCE\", \"RIOK2\", \"BCKDHB\", \"RAD1\", \"NREP\", \"ADH5\", \"PLRG1\",\n",
    "  \"ROR1\", \"RAB3B\", \"DIAPH3\", \"GNL2\", \"FGF2\", \"NMNAT2\", \"KIF20A\", \"CENPI\", \"DDX1\", \"XXYLT1\",\n",
    "  \"GPR176\", \"BBS9\", \"C14orf166\", \"BOD1\", \"CDC123\", \"SNRPD3\", \"FAM118B\", \"DPH3\", \"EIF2B3\", \"RPF2\",\n",
    "  \"APLP1\", \"DACT1\", \"PDHB\", \"C14orf119\", \"DTD1\", \"SAMM50\", \"CCL26\", \"MED20\", \"UTP6\", \"RARS2\",\n",
    "  \"ARMCX2\", \"RARS\", \"MTHFD2\", \"DHX15\", \"HTR7\", \"MTHFD1L\", \"ARMC9\", \"XPOT\", \"IARS\", \"HDX\",\n",
    "  \"ACTRT3\", \"ERCC2\", \"TBC1D16\", \"GARS\", \"KIF7\", \"UBE2K\", \"SLC25A3\", \"ICMT\", \"UGGT2\", \"ATP11C\",\n",
    "  \"SLC24A1\", \"EIF2AK4\", \"GPX8\", \"ALX1\", \"OSTC\", \"TRPC4\", \"HAS2\", \"FZD2\", \"TRNT1\", \"MMADHC\",\n",
    "  \"SNX8\", \"CDH6\", \"HAT1\", \"SEC11A\", \"DIMT1\", \"TM2D2\", \"FST\", \"GBE1\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    " Exhaustion <- c('PDCD1', 'LAYN', 'HAVCR2', 'LAG3', 'CD244', 'CTLA4', 'LILRB1', 'TIGIT', 'TOX', 'VSIR', 'BTLA', 'ENTPD1', 'CD160', 'LAIR1')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GEP1 <- c(\"UBE2S\", \"NME1\", \"TUBB2B\", \"U2AF1\", \"ATP5G1\", \"CDKN2A\", \"MRPL12\", \"FABP6\", \"PTTG1\", \"TNFAIP6\", \n",
    "\"GADD45A\", \"HLA-DQA2\", \"SF3B14\", \"NDUFB8.1\", \"NHP2L1\", \"ATP5O\", \"MRPL23\", \"CNN3\", \"ATP5B\", \"CKS1B\", \n",
    "\"CCDC28B\", \"RGS16\", \"RPS4Y1\", \"ATPIF1\", \"ATP5G3\", \"HIST1H4C\", \"TK1\", \"PSMA2.1\", \"TAGLN2\", \"H2AFZ\", \n",
    "\"TUBA1C\", \"SYNGR2\", \"BIRC5\", \"LDHB\", \"DYNLT1\", \"EIF4A1\", \"ENO1\", \"ATP5I\", \"KPNA2\", \"RPL22L1\", \n",
    "\"TMC8\", \"C19orf10\", \"TXN\", \"KRT18\", \"REL\", \"KIAA0101\", \"UBE2C\", \"RPL27\", \"PRRX1\", \"USMG5\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GEPs_Signature <- read.csv(\"Results/06.Epithelial/cNMF_Results/cNMF_topgenes_df.csv\")\n",
    "GEP1 <- GEPs_Signature$X1[1:100]\n",
    "GEP1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Meta_signature <-   list(Stem.Sig2 = Stem.Sig2,\n",
    "                         GEP1 = GEP1,\n",
    "                         SPP1_Macro = Myeloid.Markers$X12[1:50],\n",
    "                         IL1B_Macro = Myeloid.Markers$X4[1:50],\n",
    "                         C1QC_Macro = Myeloid.Markers$X2[1:50],\n",
    "                         Tex = T.Markers$X26[1:50],\n",
    "                         Tpex = T.Markers$X1[1:50],\n",
    "                         TNFRSF9_N_Treg = T.Markers$X19[1:50],\n",
    "                         TNFRSF9_P_Treg = T.Markers$X21[1:50],\n",
    "                         Exhaustion = Exhaustion,\n",
    "                         #M2_signature = Myeloid.Immunity.geneset$`M2 Macrophage Polarization`,\n",
    "                         MYC = \"MYC\",\n",
    "                         PDCD1 = \"PDCD1\",\n",
    "                         LAG3 = \"LAG3\",\n",
    "                         TOX = \"TOX\",\n",
    "                         HAVCR2 = \"HAVCR2\",\n",
    "                         CXCL13 = \"CXCL13\",\n",
    "                         MIF = \"MIF\",\n",
    "                         CD44 = \"CD44\",\n",
    "                         C1QC = \"C1QC\",\n",
    "                         IL1B = \"IL1B\",\n",
    "                         SPP1 = \"SPP1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "y=as.matrix(GSE102349_Dataset)\n",
    "geneSets=Meta_signature"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## build GSEA parameter object\n",
    "ssgseaParam <- ssgseaParam(y, geneSets)\n",
    "## estimate GSEA enrichment scores for the three sets\n",
    "gsea_es <- gsva(ssgseaParam)\n",
    "GSE102349_Scores <- as.data.frame(t(gsea_es))\n",
    "GSE102349_Scores$SampleID <- rownames(GSE102349_Scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(GSE102349_Scores, aes_string(x = \"PDCD1\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"PDCD1\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_PDCD1_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"LAG3\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"LAG3\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_LAG3_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"HAVCR2\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"HAVCR2\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_HAVCR2_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"Exhaustion\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"Exhaustion\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_Exhaustion_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"Tex\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"Tex\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_Tex_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"Tpex\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"Tpex\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_Tpex_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"TNFRSF9_N_Treg\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"TNFRSF9_N_Treg\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_TNFRSF9_N_Treg_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n",
    "\n",
    "\n",
    "\n",
    "ggplot(GSE102349_Scores, aes_string(x = \"TNFRSF9_P_Treg\", y = \"SPP1_Macro\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"TNFRSF9_P_Treg\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_TNFRSF9_P_Treg_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(GSE102349_Scores, aes_string(x = \"Stem.Sig2\", y = \"SPP1\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"Stemness Index\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_Stem.Sig2_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(GSE102349_Scores, aes_string(x = \"MIF\", y = \"SPP1\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_MIF_SPP1.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GSE102349_Sig_Clinical <-  merge(x = GSE102349_Scores, y = GSE102349_Clinical, by = \"SampleID\", all = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GSE102349_Sig_Survival <- subset(GSE102349_Sig_Clinical,Time_months!=\"N/A\")\n",
    "# bulk_survival <- bulk_clinical\n",
    "GSE102349_Sig_Survival$Time_months <- as.numeric(GSE102349_Sig_Survival$Time_months)\n",
    "GSE102349_Sig_Survival$PFS_status <- as.numeric(GSE102349_Sig_Survival$PFS_status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "Stem.Sig2.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'Stem.Sig2') \n",
    "Stem.Sig2.cut \n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(Stem.Sig2_2 = ifelse(Stem.Sig2 >= Stem.Sig2.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$Stem.Sig2_2 <- factor(GSE102349_Sig_Survival$Stem.Sig2_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~Stem.Sig2_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_Stem.Sig2 <- survfit(Surv(Time_months,PFS_status) ~ Stem.Sig2_2, data = GSE102349_Sig_Survival)\n",
    "\n",
    "ggsurv_Stem.Sig2_2 <- ggsurvplot(\n",
    "  fit_Stem.Sig2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.10),\n",
    "  pval.coord = c(18, 0.10),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25, \n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"Stemness Index\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_Stem.Sig2_2$plot <- ggsurv_Stem.Sig2_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0), \n",
    "    labels = function(x) x * 100 \n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20), \n",
    "    legend.position = c(0.05, 0.12), \n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE) \n",
    "  )\n",
    "\n",
    "ggsurv_Stem.Sig2_2$table <- ggsurv_Stem.Sig2_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "\n",
    "ggsurv_Stem.Sig2_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_Stem.Sig2.pdf\", ggsurv_Stem.Sig2_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GEP1.cut <- surv_cutpoint(GSE102349_Sig_Survival, time = 'Time_months', event = 'PFS_status', variables = 'GEP1') \n",
    "GEP1.cut  \n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>%\n",
    "  mutate(GEP1_2 = ifelse(GEP1 >= GEP1.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$GEP1_2 <- factor(GSE102349_Sig_Survival$GEP1_2, levels = c(\"Low\", \"High\"))\n",
    "\n",
    "res.cox1 <- coxph(Surv(Time_months, PFS_status) ~ GEP1_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "\n",
    "fit_GEP1 <- survfit(Surv(Time_months, PFS_status) ~ GEP1_2, data = GSE102349_Sig_Survival)\n",
    "\n",
    "ggsurv_GEP1_2 <- ggsurvplot(\n",
    "  fit_GEP1, \n",
    "  palette = \"jama\",\n",
    "  pval = TRUE,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = TRUE,\n",
    "  pval.method.coord = c(3, 0.10),\n",
    "  pval.coord = c(18, 0.10),\n",
    "  conf.int = FALSE,  \n",
    "  censor = FALSE,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = FALSE,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = FALSE, \n",
    "  risk.table.y.text.col = FALSE,\n",
    "  risk.table.y.text = FALSE,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = FALSE,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25,\n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"GEP1 Score\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_GEP1_2$plot <- ggsurv_GEP1_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0),\n",
    "    labels = function(x) x * 100\n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20),\n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0),\n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"),\n",
    "    legend.text = element_text(size = 16)\n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_GEP1_2$table <- ggsurv_GEP1_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "\n",
    "ggsurv_GEP1_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_GEP1.pdf\", ggsurv_GEP1_2$plot,\n",
    "       width = 5, height = 5, units = \"in\", dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggplot(GSE102349_Scores, aes_string(x = \"GEP1\", y = \"Stem.Sig2\")) +\n",
    "  geom_point(size = 2,color = c('#EC0101'), alpha = 0.5) +\n",
    "  theme_bw() + \n",
    "  theme(axis.title = element_text(size = 16),\n",
    "        axis.text = element_text(size = 14),\n",
    "        axis.ticks.length = unit(0.25,'cm'),\n",
    "        axis.ticks = element_line(size = 1),\n",
    "        panel.border = element_rect(size = 1.5),\n",
    "        panel.grid = element_blank()\n",
    "  ) + xlab(\"GEP1 Score\") + ylab(\"Stemness Index\") +\n",
    "  geom_smooth(method = 'lm',se = T,color = '#F9B208',size = 1.5,fill = '#FEA82F') +\n",
    "  stat_cor(method = \"pearson\",digits = 3,size=6)\n",
    "\n",
    "ggsave(\"Results/09.GEO_validation/Scatter_GEP1_Stem.Sig2.pdf\", width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "MIF.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'MIF') \n",
    "MIF.cut \n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(MIF_2 = ifelse(MIF >= MIF.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$MIF_2 <- factor(GSE102349_Sig_Survival$MIF_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~MIF_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_MIF_2 <- survfit(Surv(Time_months,PFS_status) ~ MIF_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_MIF_2 <- ggsurvplot(\n",
    "  fit_MIF_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25,\n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"MIF Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_MIF_2$plot <- ggsurv_MIF_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0), \n",
    "    labels = function(x) x * 100 \n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20),\n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_MIF_2$table <- ggsurv_MIF_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20) \n",
    "  )\n",
    "ggsurv_MIF_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_MIF_2.pdf\", ggsurv_MIF_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "SPP1.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'SPP1') \n",
    "SPP1.cut\n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(SPP1_2 = ifelse(SPP1 >= SPP1.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$SPP1_2 <- factor(GSE102349_Sig_Survival$SPP1_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~SPP1_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_SPP1_2 <- survfit(Surv(Time_months,PFS_status) ~ SPP1_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_SPP1_2 <- ggsurvplot(\n",
    "  fit_SPP1_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25, \n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"SPP1 Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_SPP1_2$plot <- ggsurv_SPP1_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0),\n",
    "    labels = function(x) x * 100\n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20),\n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0), \n",
    "    #legend.background = element_blank(),\n",
    "    #legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_SPP1_2$table <- ggsurv_SPP1_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "ggsurv_SPP1_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_SPP1_2.pdf\", ggsurv_SPP1_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "SPP1_Macro.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'SPP1_Macro') \n",
    "SPP1_Macro.cut\n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(SPP1_Macro_2 = ifelse(SPP1_Macro >= SPP1_Macro.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$SPP1_Macro_2 <- factor(GSE102349_Sig_Survival$SPP1_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~SPP1_Macro_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_SPP1_Macro_2 <- survfit(Surv(Time_months,PFS_status) ~ SPP1_Macro_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_SPP1_Macro_2 <- ggsurvplot(\n",
    "  fit_SPP1_Macro_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25,\n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"SPP1+ Macrophage Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_SPP1_Macro_2$plot <- ggsurv_SPP1_Macro_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0), \n",
    "    labels = function(x) x * 100 \n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20), \n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_SPP1_Macro_2$table <- ggsurv_SPP1_Macro_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "ggsurv_SPP1_Macro_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_SPP1_Macro_2.pdf\", ggsurv_SPP1_Macro_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "C1QC.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'C1QC') \n",
    "C1QC.cut\n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(C1QC_2 = ifelse(C1QC >= C1QC.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$C1QC_2 <- factor(GSE102349_Sig_Survival$C1QC_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~C1QC_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_C1QC <- survfit(Surv(Time_months,PFS_status) ~ C1QC_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_C1QC <- ggsurvplot(\n",
    "  fit_C1QC, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25,\n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"C1QC+ Macrophage Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_C1QC$plot <- ggsurv_C1QC$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0), \n",
    "    labels = function(x) x * 100 \n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20), \n",
    "    legend.position = c(0.05, 0.12), \n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE) \n",
    "  )\n",
    "\n",
    "ggsurv_C1QC$table <- ggsurv_C1QC$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20) \n",
    "  )\n",
    "\n",
    "ggsurv_C1QC\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_C1QC.pdf\", ggsurv_C1QC$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "C1QC_Macro.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'C1QC_Macro') \n",
    "C1QC_Macro.cut\n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(C1QC_Macro_2 = ifelse(C1QC_Macro >= C1QC_Macro.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$C1QC_Macro_2 <- factor(GSE102349_Sig_Survival$C1QC_Macro_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~C1QC_Macro_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_C1QC_Macro_2 <- survfit(Surv(Time_months,PFS_status) ~ C1QC_Macro_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_C1QC_Macro_2 <- ggsurvplot(\n",
    "  fit_C1QC_Macro_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25,\n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"C1QC+ Macrophage Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_C1QC_Macro_2$plot <- ggsurv_C1QC_Macro_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0),\n",
    "    labels = function(x) x * 100\n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20), \n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_C1QC_Macro_2$table <- ggsurv_C1QC_Macro_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20) \n",
    "  )\n",
    "ggsurv_C1QC_Macro_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_C1QC_Macro_2.pdf\", ggsurv_C1QC_Macro_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "IL1B.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'IL1B') \n",
    "IL1B.cut \n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(IL1B_2 = ifelse(IL1B >= IL1B.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$IL1B_2 <- factor(GSE102349_Sig_Survival$IL1B_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~IL1B_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_IL1B_2 <- survfit(Surv(Time_months,PFS_status) ~ IL1B_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_IL1B_2 <- ggsurvplot(\n",
    "  fit_IL1B_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25, \n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"IL1B Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "ggsurv_IL1B_2$plot <- ggsurv_IL1B_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0), \n",
    "    labels = function(x) x * 100 \n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20), \n",
    "    legend.position = c(0.05, 0.12), \n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_IL1B_2$table <- ggsurv_IL1B_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "ggsurv_IL1B_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_IL1B_2.pdf\", ggsurv_IL1B_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "IL1B_Macro.cut <- surv_cutpoint(GSE102349_Sig_Survival, time= 'Time_months', event = 'PFS_status', variables = 'IL1B_Macro') \n",
    "IL1B_Macro.cut \n",
    "GSE102349_Sig_Survival <- GSE102349_Sig_Survival %>% mutate(IL1B_Macro_2 = ifelse(IL1B_Macro >= IL1B_Macro.cut$cutpoint[[1]], \"High\", \"Low\"))   \n",
    "GSE102349_Sig_Survival$IL1B_Macro_2 <- factor(GSE102349_Sig_Survival$IL1B_Macro_2, levels = c(\"Low\",\"High\"))\n",
    "res.cox1 <- coxph(Surv(Time_months,PFS_status)~IL1B_Macro_2, data = GSE102349_Sig_Survival)\n",
    "publish(res.cox1)\n",
    "fit_IL1B_Macro_2 <- survfit(Surv(Time_months,PFS_status) ~ IL1B_Macro_2, data = GSE102349_Sig_Survival)\n",
    "ggsurv_IL1B_Macro_2 <- ggsurvplot(\n",
    "  fit_IL1B_Macro_2, \n",
    "  palette = \"jama\",\n",
    "  pval = T,\n",
    "  pval.size = 5, \n",
    "  digits = 3,\n",
    "  pval.method = T,\n",
    "  pval.method.coord = c(3, 0.1),\n",
    "  pval.coord = c(18, 0.1),\n",
    "  conf.int = F,  \n",
    "  censor = F,\n",
    "  censor.shape = \"|\", \n",
    "  censor.size = 3, \n",
    "  test.for.trend = F,\n",
    "  xlim = c(0, 50),\n",
    "  ylim = c(0, 1.05),\n",
    "  break.y.by = 0.2,\n",
    "  surv.scale = \"percent\",\n",
    "  lty = 1:1,\n",
    "  xlab = \"Time in months\", \n",
    "  ylab = \"Progression-Free Survival (%)\",\n",
    "  break.time.by = 12, \n",
    "  axes.offset = FALSE,\n",
    "  risk.table = F, \n",
    "  risk.table.y.text.col = F,\n",
    "  risk.table.y.text = F,\n",
    "  risk.table.font.tickslab.x = FALSE,\n",
    "  tables.y.text = F,\n",
    "  risk.table.pos = \"out\", \n",
    "  tables.height = 0.25, \n",
    "  font.risk.table.x = c(12, \"plain\", \"black\"),\n",
    "  font.x = c(16, \"plain\", \"black\"),\n",
    "  font.y = c(16, \"plain\", \"black\"),\n",
    "  font.tickslab = c(14, \"plain\", \"black\"),\n",
    "  legend = \"top\",            \n",
    "  legend.title = \"IL1B+ Macrophage Signature\",        \n",
    "  legend.labs = c(\"Low\", \"High\"),\n",
    "  font.legend = c(16, \"plain\", \"black\"),\n",
    "  font.legend.title = c(20, \"plain\", \"black\"),\n",
    "  font.legend.labs = c(16, \"plain\", \"black\"),\n",
    "  linetype = \"solid\",\n",
    "  tables.theme = theme_cleantable()\n",
    ")\n",
    "\n",
    "\n",
    "ggsurv_IL1B_Macro_2$plot <- ggsurv_IL1B_Macro_2$plot +\n",
    "  scale_y_continuous(\n",
    "    expand = c(0, 0),\n",
    "    labels = function(x) x * 100\n",
    "  ) +\n",
    "  theme(\n",
    "    plot.margin = margin(20, 20, 20, 20),\n",
    "    legend.position = c(0.05, 0.12),\n",
    "    legend.justification = c(0, 0), \n",
    "    legend.background = element_blank(),\n",
    "    legend.key = element_blank(),\n",
    "    legend.key.height = unit(1, \"cm\"), \n",
    "    legend.text = element_text(size = 16) \n",
    "  ) +\n",
    "  guides(\n",
    "    color = guide_legend(ncol = 2, byrow = TRUE)\n",
    "  )\n",
    "\n",
    "ggsurv_IL1B_Macro_2$table <- ggsurv_IL1B_Macro_2$table +\n",
    "  theme(\n",
    "    plot.margin = margin(0, 20, 20, 20)\n",
    "  )\n",
    "ggsurv_IL1B_Macro_2\n",
    "ggsave(\"Results/09.GEO_validation/Survival_Curves_IL1B_Macro_2.pdf\", ggsurv_IL1B_Macro_2$plot, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "### Correlation analysis between Stem.Sig1/SPP1/SPP1_Macro and tumor stage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GSE102349_Sig_Survival1 <- subset(GSE102349_Sig_Survival, GSE102349_Sig_Survival$Stage!=\"Unknown\")\n",
    "\n",
    "GSE102349_Sig_Survival1 <- GSE102349_Sig_Survival1 %>%\n",
    "  mutate(Stage2 = case_when(\n",
    "    Stage %in% c(\"I\", \"II\") ~ \"I-II\",\n",
    "    Stage == \"III\" ~ \"III-IV\",\n",
    "    Stage == \"IV\" ~ \"III-IV\",\n",
    "    TRUE ~ NA_character_ \n",
    "  ))\n",
    "\n",
    "GSE102349_Sig_Survival1$Stage2 <- factor(GSE102349_Sig_Survival1$Stage2)\n",
    "\n",
    "\n",
    "GSE102349_Sig_Survival1 <- GSE102349_Sig_Survival1 %>%\n",
    "  mutate(Stage3 = case_when(\n",
    "    Stage %in% c(\"I\", \"II\") ~ \"I-II\",\n",
    "    Stage == \"III\" ~ \"III\",\n",
    "    Stage == \"IV\" ~ \"IV\",\n",
    "    TRUE ~ NA_character_ \n",
    "  ))\n",
    "\n",
    "GSE102349_Sig_Survival1$Stage3 <- factor(GSE102349_Sig_Survival1$Stage3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggviolin_SPP1 <- ggviolin(GSE102349_Sig_Survival1,\n",
    "                     x = \"Stage3\", \n",
    "                     y = \"SPP1\",\n",
    "                     xlab = \"Clincial Stage\", \n",
    "                     ylab = \"SPP1 Signature\",\n",
    "                     color = \"Stage3\", \n",
    "                     palette = \"jama\",\n",
    "                     add = c(\"boxplot\", \"jitter\"),\n",
    "                     shape = \"Stage3\"\n",
    ") + \n",
    "  stat_compare_means(comparisons = list(c('I-II','III'), c('I-II','IV')), \n",
    "                     method = \"wilcox.test\", label = \"p.format\", label.y = c(1.2,1.3)) +\n",
    "  theme(legend.position = \"none\") \n",
    "ggviolin_SPP1\n",
    "ggsave(\"Results/09.GEO_validation/ggviolin_SPP1_stage.pdf\", ggviolin_SPP1, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "ggviolin_Stem.Sig2 <- ggviolin(GSE102349_Sig_Survival1,\n",
    "                     x = \"Stage3\", \n",
    "                     y = \"Stem.Sig2\",\n",
    "                     xlab = \"Clinical Stage\",\n",
    "                     ylab = \"Stemness Index\",\n",
    "                     color = \"Stage3\", \n",
    "                     palette = \"jama\",\n",
    "                     add = c(\"boxplot\", \"jitter\"),\n",
    "                     shape = \"Stage3\"\n",
    ") + \n",
    "  stat_compare_means(comparisons = list(c('I-II','III'), c('I-II','IV')), \n",
    "                     method = \"wilcox.test\", label = \"p.format\", label.y = c(0.4,0.42)) +\n",
    "  theme(legend.position = \"none\") \n",
    "ggviolin_Stem.Sig2\n",
    "ggsave(\"Results/09.GEO_validation/ggviolin_Stem.Sig2_satge.pdf\", ggviolin_Stem.Sig2, width = 5, height = 5, units = c(\"in\"), dpi = 300, limitsize = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "sessionInfo()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "NPC_Renv",
   "language": "R",
   "name": "npc_renv"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
