{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "69a28252",
   "metadata": {},
   "source": [
    "## 10_1. Preprocess, integration and clustering of NPC Spatial Data\n",
    "\n",
    "<div style=\"text-align: left;\">\n",
    "    <p style=\"text-align: left;\">Updated Time: 2025-04-09</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d67b4811-bdfe-4ec6-b3a9-ff69757523c1",
   "metadata": {},
   "source": [
    "STAligner is designed for alignment and integration of spatially resolved transcriptomics data.\n",
    "\n",
    "STAligner first normalizes the expression proï¬les for all spots and constructs a spatial neighbor network using the spatial coordinates. STAligner further employs a graph attention auto-encoder neural network to extract spatially aware embedding, and constructs the spot triplets based on current embeddings to guide the alignment process by attracting similar spots and discriminating dissimilar spots across slices. STAligner introduces the triplet loss to update the spot embedding to reduce the distance from the anchor to positive spot, and increase the distance from the anchor to negative spot. The triplet construction and auto-encoder training are optimized iteratively until batch-corrected embeddings are generated. b. STAligner can be applied to integrate ST datasets to achieve alignment and simultaneous identification of spatial domains from different biological samples in (a), technological platforms (I), developmental (embryonic) stages (II), disease conditions (III) and consecutive slices of a tissue for 3D slice alignment (IV).\n",
    "\n",
    "Zhou, X., Dong, K. & Zhang, S. Integrating spatial transcriptomics data across different conditions, technologies and developmental stages. Nat Comput Sci 3, 894â€“906 (2023). https://doi.org/10.1038/s43588-023-00528-w"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28fac73d",
   "metadata": {},
   "source": [
    "#### Loading packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1bb909d-607e-43da-bf39-d6f55faec81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import omicverse as ov\n",
    "import scanpy as sc\n",
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from pathlib import Path\n",
    "from scipy.sparse import csr_matrix\n",
    "from matplotlib.patches import Patch\n",
    "\n",
    "ov.utils.ov_plot_set()\n",
    "\n",
    "import warnings\n",
    "warnings.simplefilter(\"ignore\") "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd6064d6",
   "metadata": {},
   "source": [
    "##### Set working directory for analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9fb6e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "cwd = '/media/bio/Disk/Research Data/EBV/omicverse'\n",
    "os.chdir(cwd)\n",
    "updated_dir = os.getcwd()\n",
    "print(\"Updated working directory: \", updated_dir)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b43862-012f-4a1a-b63c-035aa4d8bc68",
   "metadata": {},
   "source": [
    "# Preprocess data\n",
    "\n",
    "Here, We use the mouse olfactory bulb data generated by Stereo-seq and Slide-seqV2. The processed Stereo-seq and Slide-seqV2 data can be downloaded from https://drive.google.com/drive/folders/1Omte1adVFzyRDw7VloOAQYwtv_NjdWcG?usp=share_link. and the original tutorals can be finded from https://staligner.readthedocs.io/en/latest\n",
    "\n",
    "Here is a critical point that must be clarified: for STAligner, it first calculates highly variable genes before concating annadata samples. Therefore, the number of highly variable genes should not be selected too low. Otherwise, in the case of a large number of samples, the downstream features for STAligner training would be insufficient, impacting the model's performance.\n",
    "\n",
    "When using STAligner, it is necessary to adjust the **rad_cutoff** parameter according to different data to ensure that each spot has an **average of 5-10 adjacent spots** connected to it. Such as: \"11.3356 neighbors per cell on average.\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06c6baa6",
   "metadata": {},
   "source": [
    "#### Preprocess NPC Spatial Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b54fd956",
   "metadata": {},
   "outputs": [],
   "source": [
    "# **Root directory for 10X Visium data**\n",
    "pathway = \"Dataset/GSE206245\"\n",
    "\n",
    "# **Batch_list to store all AnnData objects**\n",
    "Batch_list = []\n",
    "adj_list = []\n",
    "\n",
    "sample_ids = ['NPC_ST05', 'NPC_ST06', 'NPC_ST07', 'NPC_ST08', 'NPC_ST09', 'NPC_ST10', 'NPC_ST11', 'NPC_ST12', 'NPC_ST16', 'NPC_ST17', 'NPC_ST18', 'NPC_ST19']\n",
    "print(sample_ids)\n",
    "\n",
    "# **Iterate over all sample directories**\n",
    "for sample_id in sample_ids:\n",
    "    print(f\"\\nðŸŸ¢ Processing sample: {sample_id}\")\n",
    "\n",
    "    try:\n",
    "        # **Read 10X Visium data**\n",
    "        sample_path = os.path.join(pathway,sample_id)\n",
    "        adata = sc.read_visium(sample_path, library_id=sample_id)\n",
    "\n",
    "        # check whether the adata.X is sparse matrix\n",
    "        if isinstance(adata.X, pd.DataFrame):\n",
    "            adata.X = csr_matrix(adata.X)\n",
    "        else:\n",
    "            pass\n",
    "\n",
    "        # add batch name\n",
    "        adata.obs[\"sample_id\"] = sample_id\n",
    "              \n",
    "        # make var name unique\n",
    "        adata.var_names_make_unique(join=\"++\")\n",
    "\n",
    "        # make spot name unique\n",
    "        adata.obs_names = [x+'_'+sample_id for x in adata.obs_names]\n",
    "\n",
    "        # Spot filtering\n",
    "        sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "        adata = adata[adata.obs[\"total_counts\"] > 500, :] # Remove spots with low UMI counts\n",
    "        adata = adata[adata.obs[\"total_counts\"] < 60000, :] # Remove spots with extremely high UMI counts \n",
    "        adata = adata[adata.obs[\"n_genes_by_counts\"] > 100, :] # Remove spots with a low number of detected genes\n",
    "        adata = adata[adata.obs[\"n_genes_by_counts\"] < 10000, :] # Remove spots with a extremely high number of detected genes\n",
    "        adata = adata[adata.obs[\"in_tissue\"] == 1, :] # Retain only spots within the tissue (specific to Visium)\n",
    "        \n",
    "        # save adata.raw\n",
    "        adata.raw = adata.copy()\n",
    "\n",
    "        # Constructing the spatial network\n",
    "        ov.space.Cal_Spatial_Net(adata, rad_cutoff=200) # rad_cutoff need to be test, the spatial network are saved in adata.uns[â€˜adjâ€™]\n",
    "\n",
    "        # Normalization\n",
    "        sc.pp.highly_variable_genes(adata, flavor=\"seurat_v3\", n_top_genes=13000)\n",
    "        sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "        sc.pp.log1p(adata)\n",
    "        adata = adata[:, adata.var['highly_variable']]\n",
    "\n",
    "        # **Store in Batch_list**\n",
    "        adj_list.append(adata.uns['adj'])\n",
    "        Batch_list.append(adata)\n",
    "\n",
    "        print(f\"âœ… Successfully loaded: {sample_id} â†’ Dimensions: {adata.shape}\")\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"âŒ Failed to load: {sample_id}, Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f180a699-00ee-47aa-af0d-0a2c67b9c024",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat = ad.concat(Batch_list, label=\"library_id\", keys=sample_ids, uns_merge=\"unique\")\n",
    "adata_concat.obs[\"batch\"] = adata_concat.obs[\"sample_id\"].astype('category')\n",
    "adata_concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7468f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata_concat.X.shape)\n",
    "print(np.min(adata_concat.X), np.max(adata_concat.X))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f13013ce-ebd1-49da-97b3-0a7a54c4d358",
   "metadata": {},
   "source": [
    "# Training STAligner model\n",
    "\n",
    "Here, we used `ov.space.pySTAligner` to construct a STAGATE object to train the model.\n",
    "\n",
    "We are using the `train_STAligner_subgraph` function from STAligner to reduce GPU memory usage, each slice is considered as a subgraph for training."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ad357e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import numpy as np\n",
    "import torch\n",
    "\n",
    "# Set random seed for Python's built-in random module\n",
    "random.seed(666)\n",
    "\n",
    "# Set random seed for NumPy\n",
    "np.random.seed(666)\n",
    "\n",
    "# Set random seed for PyTorch (both CPU and GPU)\n",
    "torch.manual_seed(666)\n",
    "torch.cuda.manual_seed_all(666)\n",
    "\n",
    "# Ensure deterministic behavior for CUDA operations (may reduce performance)\n",
    "torch.backends.cudnn.deterministic = True\n",
    "torch.backends.cudnn.benchmark = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ec986bc-1550-41d1-9ac4-5cc5f4aedf08",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "# iter_comb is used to specify the order of integration. For example, (0, 1) means slice 0 will be algined with slice 1 as reference.\n",
    "iter_comb = [(i, i + 1) for i in range(len(sample_ids) - 1)]\n",
    "\n",
    "# Here, to reduce GPU memory usage, each slice is considered as a subgraph for training.\n",
    "STAligner_obj = ov.space.pySTAligner(adata_concat, verbose=True, knn_neigh = 100, n_epochs = 600, iter_comb = iter_comb,\n",
    "                                     batch_key = 'batch',  key_added='STAligner', Batch_list = Batch_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb84f8be-6d4a-44df-9f10-894dcb854124",
   "metadata": {},
   "outputs": [],
   "source": [
    "STAligner_obj.train()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b76d42f3-fb4a-4a54-833f-1208972297cf",
   "metadata": {},
   "source": [
    "We stored the latent embedding in `adata.obsm['STAligner']`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68ff9473-70c9-4930-adb1-a607f8c5df7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata = STAligner_obj.predicted()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a272914e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(adata.X.shape)\n",
    "print(np.min(adata.X), np.max(adata.X))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7c8f40f-36c0-4f20-b969-3c83cc249c02",
   "metadata": {},
   "source": [
    "# Clustering the space\n",
    "\n",
    "We can use `GMM`, `leiden` or `louvain` to cluster the space.\n",
    "\n",
    "`ov.utils.cluster(adata,use_rep='STAligner',method='GMM',n_components=7,covariance_type='full', tol=1e-9, max_iter=1000, random_state=3607`\n",
    "\n",
    "or `sc.pp.neighbors(adata, use_rep='STAligner', random_state=666)`            \n",
    "`ov.utils.cluster(adata,use_rSTAlignerGATE',method='leiden',resolution=1)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8cf254c2-9ad0-4783-8310-af8f60228d2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata, use_rep='STAligner', random_state=666)\n",
    "sc.tl.umap(adata_concat, random_state=666)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "df01f30d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run leiden clustering for different resolutions\n",
    "for resolution in [0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.5]:\n",
    "    ov.pp.leiden(\n",
    "        adata,\n",
    "        resolution=resolution,\n",
    "        key_added=f\"leiden_{str(resolution).replace('.', '_')}\",\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be1aab19",
   "metadata": {},
   "source": [
    "#### Plot the clustree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "06fe0cb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyclustree import clustree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f43a7c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot the clustree\n",
    "fig = clustree(\n",
    "    adata,\n",
    "    [f\"leiden_{str(resolution).replace('.', '_')}\" for resolution in [0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.5]],\n",
    "    title=\"Clustree of Spatial Niches\",\n",
    "    edge_weight_threshold=0.00,  # the minimum fraction of the parent cluster assigned to the child cluster to plot\n",
    "    show_fraction=True,  # show the fraction of cells in each cluster\n",
    ")\n",
    "fig.set_size_inches(10, 8)\n",
    "fig.set_dpi(100)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad999b04",
   "metadata": {},
   "source": [
    "#### Adding cluster scoring"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c652dccb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obsm['X_pca']=adata.obsm['STAligner']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70ff0962",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supported are Silhouette score, Calinski and Harabasz score and Davies-Bouldin score.\n",
    "fig = clustree(\n",
    "    adata,\n",
    "    [f\"leiden_{str(resolution).replace('.', '_')}\" for resolution in [0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.5]],\n",
    "    title=\"Clustree of Spatial Niches with Silhouette Score\",\n",
    "    score_clustering=\"silhouette\",\n",
    "    score_basis=\"pca\",\n",
    ")\n",
    "fig.set_size_inches(10, 8)\n",
    "fig.set_dpi(100)\n",
    "\n",
    "# Save and Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Silhouette_Score_Spatial_scNiche.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c395423a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supported are Silhouette score, Calinski and Harabasz score and Davies-Bouldin score.\n",
    "fig = clustree(\n",
    "    adata,\n",
    "    [f\"leiden_{str(resolution).replace('.', '_')}\" for resolution in [0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.5]],\n",
    "    title=\"Clustree of Spatial Niches with Calinski Harabasz Score\",\n",
    "    score_clustering=\"calinski_harabasz\",\n",
    "    score_basis=\"pca\",\n",
    ")\n",
    "fig.set_size_inches(10, 8)\n",
    "fig.set_dpi(100)\n",
    "\n",
    "# Save and Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Calinski_Harabasz_Score_Spatial_scNiche.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "32f9d359",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Supported are Silhouette score, Calinski and Harabasz score and Davies-Bouldin score.\n",
    "\n",
    "fig = clustree(\n",
    "    adata,\n",
    "    [f\"leiden_{str(resolution).replace('.', '_')}\" for resolution in [0.1, 0.2, 0.4, 0.6, 0.8, 1.0, 1.2, 1.5]],\n",
    "    title=\"Clustree of Spatial Niches with Davies Bouldin Score\",\n",
    "    score_clustering=\"davies_bouldin\",\n",
    "    score_basis=\"pca\",\n",
    ")\n",
    "fig.set_size_inches(10, 8)\n",
    "fig.set_dpi(100)\n",
    "\n",
    "# Save and Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Davies_Bouldin_Score_Spatial_scNiche.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "596f9598",
   "metadata": {},
   "source": [
    "Based on the cluster scoringï¼Œa resolution of 0.2 may be the optimal. Here we visualize the optimal clustering using UMAP representation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7204cf95",
   "metadata": {},
   "outputs": [],
   "source": [
    "original_clusters = sorted(adata_concat.obs['leiden_0_2'].unique(), key=lambda x: int(x))\n",
    "new_labels = [f\"Niche{i+1}\" for i in range(len(original_clusters))]\n",
    "label_map = dict(zip(original_clusters, new_labels))\n",
    "adata_concat.obs['scNiche'] = adata_concat.obs['leiden_0_2'].map(label_map)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f889185",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import patheffects\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "ov.pl.embedding(adata_concat,\n",
    "                  basis='X_umap',\n",
    "                  color=['scNiche'], \n",
    "                  palette='Paired',\n",
    "                  show=False, legend_loc=None, add_outline=False, \n",
    "                  frameon='small',legend_fontoutline=2,ax=ax\n",
    "                 )\n",
    "\n",
    "ov.utils.gen_mpl_labels(\n",
    "    adata_concat,\n",
    "    'scNiche',\n",
    "    exclude=(\"None\",),  \n",
    "    basis='X_umap',\n",
    "    ax=ax,\n",
    "    adjust_kwargs=dict(arrowprops=dict(arrowstyle='-', color='black')),\n",
    "    text_kwargs=dict(fontsize= 9,weight='bold',\n",
    "                     path_effects=[patheffects.withStroke(linewidth=2, foreground='w')] ),\n",
    ")\n",
    "\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Umap_Spatial_scNiche.pdf\", format='pdf')\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79ebe51a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from matplotlib import patheffects\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(4,4))\n",
    "ov.pl.embedding(adata_concat,\n",
    "                  basis='X_umap',\n",
    "                  color=['batch'], \n",
    "                  palette='Paired',\n",
    "                  show=False, \n",
    "                  legend_fontoutline=2,\n",
    "                  title='',\n",
    "                  ax=ax\n",
    "                 )\n",
    "\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Umap_Batch_Spatial.pdf\", format='pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22a17f0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig,ax=plt.subplots(figsize = (6,4))\n",
    "ov.pl.cellproportion(adata=adata_concat, celltype_clusters='scNiche', groupby='batch', legend=True, ax=ax)\n",
    "# Save and Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Composition_Spatial_scNiche.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bdc79e5a",
   "metadata": {},
   "source": [
    "We can also map the clustering results back to the original spatial coordinates to obtain spatially specific clustering results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3325378",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get all batch names\n",
    "sample_ids = adata_concat.obs['sample_id'].unique()\n",
    "\n",
    "# Define grid layout for subplots\n",
    "n_cols = 4\n",
    "n_rows = -(-len(sample_ids) // n_cols)  # Ceiling division\n",
    "\n",
    "# Create figure and subplots\n",
    "fig, axes = plt.subplots(\n",
    "    n_rows, n_cols, figsize=(4 * n_cols, 4 * n_rows),\n",
    "    gridspec_kw={'wspace': 0.1, 'hspace': 0.1}\n",
    ")\n",
    "axes = axes.flatten()\n",
    "\n",
    "title_size = 12\n",
    "\n",
    "# Extract global Niche categories and corresponding colors\n",
    "# Sort Niche labels by numeric order\n",
    "Niche_labels = sorted(adata_concat.obs['scNiche'].unique(), key=lambda x: int(x.replace(\"Niche\", \"\")))\n",
    "Niche_colors = adata_concat.uns['scNiche_colors']\n",
    "legend_data = dict(zip(Niche_labels, Niche_colors))\n",
    "\n",
    "# Plot spatial clusters for each sample_id\n",
    "for i, sample_id in enumerate(sample_ids):\n",
    "    adata_subset = adata_concat[adata_concat.obs['sample_id'] == sample_id]\n",
    "\n",
    "    # Plot spatial cluster map\n",
    "    sc.pl.embedding(\n",
    "        adata_subset,\n",
    "        color='scNiche',\n",
    "        title=sample_id,\n",
    "        basis=\"spatial\",\n",
    "        alpha=0.75,\n",
    "        legend_fontsize=12,\n",
    "        legend_loc=None,  # Disable individual subplot legends\n",
    "        show=False,\n",
    "        ax=axes[i]\n",
    "    )\n",
    "\n",
    "    # Set title, invert Y axis, and hide axis ticks\n",
    "    ax = axes[i]\n",
    "    ax.set_title(sample_id, size=title_size)\n",
    "    ax.invert_yaxis()\n",
    "    ax.set_xticks([])  # Hide x-axis ticks\n",
    "    ax.set_yticks([])  # Hide y-axis ticks\n",
    "    ax.set_xlabel(\"\")\n",
    "    ax.set_ylabel(\"\")\n",
    "\n",
    "\n",
    "# Remove empty subplots (if any)\n",
    "for j in range(i + 1, len(axes)):\n",
    "    fig.delaxes(axes[j])\n",
    "\n",
    "# Rebuild unified legend (ensure sorted)\n",
    "legend_data = dict(zip(Niche_labels, Niche_colors))\n",
    "handles = [Patch(facecolor=color, label=label) for label, color in legend_data.items()]\n",
    "\n",
    "fig.legend(\n",
    "    handles=handles,\n",
    "    loc='center left',\n",
    "    bbox_to_anchor=(0.98, 0.5),\n",
    "    fontsize=14,\n",
    "    title='scNiche',\n",
    "    frameon=False,\n",
    "    ncol=1\n",
    ")\n",
    "\n",
    "# Adjust layout to make space for the right legend\n",
    "plt.tight_layout()\n",
    "\n",
    "# Save and Show plot\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"Results/10.NPC_ST_Analysis/Combined_Spatial_scNiche.pdf\", bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1fd391e",
   "metadata": {},
   "source": [
    "#### Save Spatial AnnData object with clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97bc5770",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a080bf49",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat = adata_concat.raw.to_adata()\n",
    "print(adata_concat.X.shape)\n",
    "print(np.min(adata_concat.X), np.max(adata_concat.X))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ca118d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04ea314b",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_concat.obs.to_csv(\"Processed Data/scNiche_metadata.csv\", index=True)\n",
    "adata_concat.write('Processed Data/GSE206245_NPC_ST_Cluster.h5ad',compression='gzip')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66a09702",
   "metadata": {},
   "source": [
    "\n",
    "**<span style=\"font-size:16px;\">Session informationï¼š</span>**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbe74645",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import platform\n",
    "import pkg_resources\n",
    "\n",
    "# Get Python version information\n",
    "python_version = sys.version\n",
    "# Get operating system information\n",
    "os_info = platform.platform()\n",
    "# Get system architecture information\n",
    "architecture = platform.architecture()[0]\n",
    "# Get CPU information\n",
    "cpu_info = platform.processor()\n",
    "# Print Session information\n",
    "print(\"Python version:\", python_version)\n",
    "print(\"Operating system:\", os_info)\n",
    "print(\"System architecture:\", architecture)\n",
    "print(\"CPU info:\", cpu_info)\n",
    "\n",
    "# Print imported packages and their versions\n",
    "print(\"\\nImported packages and their versions:\")\n",
    "for package in pkg_resources.working_set:\n",
    "    print(package.key, package.version)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "npc_env",
   "language": "python",
   "name": "npc_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
